craft_application/util/system.py

```diff
@@ -0,0 +1,105 @@
+# This file is part of craft-application.
+#
+# Copyright 2024 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify it
+# under the terms of the GNU Lesser General Public License version 3, as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranties of MERCHANTABILITY,
+# SATISFACTORY QUALITY, or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License along
+# with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""System-level util functions."""
+from __future__ import annotations
+
+import os
+
+from craft_cli import emit
+
+from craft_application.errors import InvalidParameterError
+
+
+def _verify_parallel_build_count(env_name: str, parallel_build_count: int | str) -> int:
+    """Verify the parallel build count is valid.
+
+    :param env_name: The name of the environment variable being checked.
+    :param parallel_build_count: The value of the variable.
+    :return: The parallel build count as an integer.
+    """
+    try:
+        parallel_build_count = int(parallel_build_count)
+    except ValueError as err:
+        raise InvalidParameterError(env_name, str(os.environ[env_name])) from err
+
+    # Ensure the value is valid positive integer
+    if parallel_build_count < 1:
+        raise InvalidParameterError(env_name, str(parallel_build_count))
+
+    return parallel_build_count
+
+
+def get_parallel_build_count(app_name: str) -> int:
+    """Get the number of parallel builds to run.
+
+    The parallel build count is determined by the first available of the
+    following environment variables in the order:
+
+    - <APP_NAME>_PARALLEL_BUILD_COUNT
+    - CRAFT_PARALLEL_BUILD_COUNT
+    - <APP_NAME>_MAX_PARALLEL_BUILD_COUNT
+    - CRAFT_MAX_PARALLEL_BUILD_COUNT
+
+    where the MAX_PARALLEL_BUILD_COUNT variables are dynamically compared to
+    the number of CPUs, and the smaller of the two is used.
+
+    If no environment variable is set, the CPU count is used.
+    If the CPU count is not available for some reason, 1 is used as a fallback.
+    """
+    parallel_build_count = None
+
+    # fixed parallel build count environment variable
+    for env_name in [
```

Comment:
This seems like something we could use the configuration service for?

Reply:
Sure does, but that seems a bit outside the scope here perhaps?  This is just code I moved from lifecycle.py.

Reply:
I'm not clear why that move was needed though, and one thing it does is prevents us from using the configuration service later. If it was done to enable getting the build count in the when the application creates a ProjectInfo in `_expand_environment`, that doesn't AFAICT get used during the actual lifecycle run, so we can leave this attached to the lifecycle service. If it somehow does, just make the method public.

Reply:
Interesting, thanks! In that case let's make `get_parallel_build_count` a public method on the LifecycleService since these are specific to the lifecycle.

Reply:
We can make `get_parallel_build_count` a public method on the lifecycle service, but then we need to call it to instantiate ProjectInfo, which instantiates the lifecycle service too early, causing the errors I pasted above.  So this code needs to be moved out of the lifecycle service.

Reply:
Hmmm that's unfortunate. I'd like to fix that later, but that'll be a pretty major change.